tools/rewriter/GenCallAsm.cpp

```diff
@@ -954,6 +976,62 @@ static void emit_return(AsmWriter &aw) {
   add_asm_line(aw, "ret");
 }
 
+/// Check types that have (both) {con,de}structors.
+/// Check types in the type registry.
+/// If the function is itself a constructor or destructor,
+/// call `ia2_type_registry_{con,de}struct`.
+/// Otherwise, go through all of the args,
+/// and if they have (both) {con,de}structors
+/// and are one of the 6 x86 register args,
+/// call `ia2_type_registry_check`.
+static void emit_type_registry_checks(Context &ctx, FnSignature sig, const std::optional<std::string> target_name, Arch arch, AsmWriter &aw) {
+  // TODO make x86 only
+  std::string_view ia2_type_registry_fn_kind;
+  std::string ia2_type_registry_fn_name;
+  std::vector<size_t> arg_indices_to_check;
+  if (target_name) {
+    if (ctx.constructors.find(*target_name) != ctx.constructors.end()) {
+      ia2_type_registry_fn_kind = "construct";
+      arg_indices_to_check.emplace_back(0);
+    } else if (ctx.destructors.find(*target_name) != ctx.destructors.end()) {
+      ia2_type_registry_fn_kind = "destruct";
+      arg_indices_to_check.emplace_back(0);
+    } else {
+      ia2_type_registry_fn_kind = "check";
+      for (auto i = 0; i < sig.api.args.size(); i++) {
+        const auto &arg = sig.api.args[i];
+        const auto &type = ctx.types.get(arg.type);
+        if (!type.has_structors()) {
+          continue;
+        }
+        if (i >= x86_int_param_reg_order.size()) {
+          llvm::errs() << "skipping checking type " << type.canonical_name << " type ID " << type.id << " for arg #" << i << " of function " << *target_name << "\n";
+          continue;
+        }
+        arg_indices_to_check.emplace_back(i);
+      }
+    }
+    ia2_type_registry_fn_name = llvm::formatv("ia2_type_registry_{0}", ia2_type_registry_fn_kind);
+  }
+
+  if (!arg_indices_to_check.empty()) {
+    emit_save_args(aw, arch);
+    emit_align_stack_8_to_16(aw, arch);
+    for (const auto i : arg_indices_to_check) {
+      const auto &arg = sig.api.args[i];
+      const auto &type = ctx.types.get(arg.type);
+      add_comment_line(aw, llvm::formatv("{0}ing type {1}, type ID {2}, arg #{3}", ia2_type_registry_fn_kind, type.canonical_name, type.id, i));
+      // `ia2_type_registry_*` arg 0 is a ptr of type `T*` from arg i.
+      add_asm_line(aw, llvm::formatv("movq {0}(%rsp), %{1}", 8 + 8 * (x86_int_param_reg_order.size() - 1 - i), x86_int_param_reg_order[0]));
+      // `ia2_type_registry_*` arg 1 is the `TypeId`.
+      add_asm_line(aw, llvm::formatv("movq ${0}, %{1}", type.id, x86_int_param_reg_order[1]));
+      emit_direct_call(aw, arch, ia2_type_registry_fn_name);
```

Comment:
I think the segfault I'm seeing is a compartment violation in the type registry library.  Do I need to declare the globals in the type registry as `IA2_SHARED_DATA`?

Reply:
So it seems #578 caused this.  At least, previously (and with #578 reverted) the tests passed locally and in most CI runs (I think one or two combinations still failed), but now it fails in all of them.

Reply:
Logically speaking, we shouldn't be linking libia2 into call gates, so I'd rather we not revert #578 wholesale. But the type registry itself may need to be linked into call-gates that use it, because logically it is part of the compartmentalized library (we want to avoid other compartments corrupting this one's idea of which objects exist and what types they have). This seems related to #526--essentially, validation of inputs to a compartment is up to an individual compartment rather than IA2 itself, and some compartments might use the type registry while others might have other means of data validation.